thirdparty/libart/src/art.c

```diff
@@ -21,6 +33,44 @@ static uint32_t __inline __builtin_ctz(uint32_t x) {
 }
 #endif
 
+// The ART implementation requires that no key be a full prefix of an existing
+// tree during insertion.  In practice this means that each key must have a
+// terminator character.  One approach is to ensure that the key and key_len
+// includes a physical trailing NUL terminator when inserting C-strings.
+// This doesn't help a great deal when working with binary strings that may be
+// a slice in the middle of a buffer that has no termination.
+//
+// To facilitate this the key_at() function is used to look up the byte
+// value at a given index.  If that index is 1 byte after the end of the
+// key, we synthesize a fake NUL terminator byte.
```

Comment:
I'd add a note here that this still assumes that the binary strings do not have embedded null bytes (which is valid because we only insert paths, and paths can never contain null bytes). If the binary strings did have embedded NULs then this approach wouldn't work.